### DISCUSSION TITLE
What's the best way to obtain an IStorageProvider in a xplat way?

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#10227] [Q&A] [answered]
I just upgraded to the 11.0.0-preview4 version and switched from the old file dialogs to IStorageProvider. I am also working on getting my app to run in Android where before it was only on Windows. 

Previous to this, I had wired up my view and view model to allow the view model code to activate the file dialogs with this code in my UserControl-derived class:

```cs
this.WhenActivated(d => d(this.ViewModel!.GameFilePicker.RegisterHandler(ctxt => FilePicker.PickFileAsync(ctxt, (Window)this.VisualRoot!))));
```

I needed the `Window` to supply the window owner of the popup dialog. 
But in the new world, I need the Window (or a `TopLevel` I guess) for its `IStorageProvider`. The cast to `Window` fails on Android (which makes sense). `VisualRoot` is only typed to be a `IRenderRoot`, so what's the best way for xplat UserControl code to obtain an `IStorageProvider`?

**Comment by AArnott** [comment, marked as answer]
It looks like this will work:

```cs
internal static IStorageProvider GetStorageProvider(this Visual visual) => 
    ((TopLevel)(visual.GetVisualRoot() ?? throw new ArgumentException("No root available"))).StorageProvider;
```

But I don't know if that's the best/safest way.

Upd:
Or in preview5+:
```
var storageProvider = TopLevel.GetTopLevel(control).StorageProvider;
```

**Reply by maxkatz6** [reply]
You can simplify it to `TopLevel.GetTopLevel(control).StorageProvider`

**Reply by AArnott** [reply]
I don't see that API:
![image](https://user-images.githubusercontent.com/3548/216841507-263d1f65-0d80-4c56-811c-8e35019edf65.png)
Is it added in something later than 11.0.0-preview4?

**Reply by maxkatz6** [reply]
Yes, it's part of preview5.

**Reply by AArnott** [reply]
Got it. If you post this as an independent answer, I'll mark it as the accepted one.

**Reply by maxkatz6** [reply]
It's fine. I updated and marked current thread.
